File: week3/example10/fit_example_10.py
import math

from ROOT import TFile, TH1D, TCanvas, gStyle, gROOT, gPad, TF1

gROOT.SetBatch(True)

def background(x, par):
    return par[0] + par[1]*x[0]
    
def gaus_peak(x, par):
    arg = 0
    if par[2] != 0:
        arg = (x[0] - par[1]) / par[2]
    func_val = par[0] * math.exp(-0.5*arg*arg)
    return func_val
    
def comb_func(x, par):
    return background(x, par) + gaus_peak(x, (par[2], par[3], par[4]))

def fit_example_10():
    root_file = TFile.Open("Test_Fitting_10.root")

    h_test_1 = root_file.Get("h_test_1")

    test_canvas = TCanvas("TestCanvas", "Fitting Test 10", 0, 0, 650, 450)

    backgnd = TF1("backgnd", background, 0, 200, 2)
    peak = TF1("peak", gaus_peak, 0, 200, 3)
    
    h_test_1.Fit("backgnd")

    test_canvas.Update()
    test_canvas.SaveAs("Fit_Example_10.png")

    fit = h_test_1.GetFunction("backgnd")
    chi2 = fit.GetChisquare()
    p = fit.GetParameters()
    e = [fit.GetParError(i) for i in [0, 1]]
    ndf = fit.GetNDF()

    print "\n"
    print " Output of Fit Parameters "
    print " ======================== "
    print " The chi^2 = {0} for {1} degrees of freedom".format(chi2, ndf)
    print " The fitted constant value = {0} +/- {1}".format(p[0], e[0])
    print " The fitted slope   value = {0} +/- {1}".format(p[1], e[1])
    
    #raw_input("Press enter to continue")
    
    comb = TF1("comb", comb_func, 0, 200, 5)
    
    #h_test_1.Fit("comb")
    #h_test_1.Draw()
    #test_canvas.Update()
        
    comb.SetParameter(0, p[0])
    comb.SetParameter(1, p[1])
    comb.SetParameter(2, 1000) # normalization
    comb.SetParameter(3, 50) # mean
    comb.SetParameter(4, 5) # width

    gPad.SetLogy()
    
    h_test_1.Fit("comb")
    h_test_1.Draw("same")
    
    test_canvas.Update()
    
    chi2 = comb.GetChisquare()
    p_n = comb.GetParameters()
    e_n = [comb.GetParError(i) for i in range(0, 5)]
    ndf = comb.GetNDF()
    
    peak.SetParameters(p_n[2], p_n[3], p_n[4])
    peak.SetLineColor(3)
    peak.Draw("same")
    
    test_canvas.SaveAs("Fit_Example_10_2.png")


fit_example_10()

File: week3/example10/Fit_Example_10.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>
#include <sstream>
#include <string>
#include <iostream>
#include <iomanip>
#include <fstream>

#include "TMath.h"
#include "TFile.h"
#include "TTree.h"
#include "TBrowser.h"
#include "TH1.h"
#include "TH2.h"
#include "TROOT.h"
#include "TCanvas.h"
#include "TStyle.h"
#include "TF1.h"
#include "Functions.h"

void FitExample10(){
  gROOT->Reset();
  
  // This is a comment
  
  // I am going to load a library of functions
  
  gROOT->LoadMacro("Functions.C+g");
  
  // I open the file of interest.
  
  TFile *hfile = new TFile("Test_Fitting_10.root","old","Exercise 1");
  
  // Copy the histogram into memory
  
  TH1F *h_test_1 = (TH1F*)hfile->Get("h_test_1");
  
  // Create a window to plot the rsults of the fit in
  
  TCanvas *TestCanvas = new TCanvas ("TestCanvas","Fitting Test 10",0,0,1200,800);
  
  // We now want to crate a root function based on one of the library functions that 
  // I ave created in the file Functions.C
  
  
  // the background will be over the range 0.->200., with two parameters
  
  TF1 *backgnd = new TF1("backgnd",background,0,200.,2);
  
  // The signal will also be over the range 0.->200., with 3 parameters
  
  TF1 *GPeak  = new TF1("GPeak",GausPeak,0,200.,3);
  
  // fit to the background
  
  h_test_1->Fit("backgnd");
  h_test_1->Draw("e1, same");
  
  // get out the parameters 
  
  // Get fit parameters into variables
  TF1 *fit = h_test_1->GetFunction("backgnd");
  Double_t chi2 = fit->GetChisquare();
  Double_t p[2];
  fit->GetParameters(p);
  Double_t e[2];
  for(Int_t i=0; i<2; i++){
    e[i] = fit->GetParError(i);
  }
  
  Int_t    ndf = fit->GetNDF();
  cout << endl << endl;
  cout << " Output of Fit Parameters " << endl;
  cout << " ======================== " << endl << endl;   
  
  cout << " The chi^2 = " << chi2 << " for " << ndf << " degrees of freedom " << endl;
  
  cout << " The fitted constant value = " << p[0] << " +/- " << e[0] << endl;
  cout << " The fitted slope    value = " << p[1] << " +/- " << e[1] << endl;
  
  
  
  TestCanvas->Update();
  
  
  TF1 *comb = new TF1("comb",CombFunc,0,200.,5);
  
  
  
  char *s = new char[1];
  cout << endl << " Hit <enter> to continue" << endl;
  gets(s);
  
  //Try again by setting some initial parameters
  
  comb->SetParameter(0,p[0]);
  comb->SetParameter(1,p[1]);
  
  // set Gaussian initial parameters
  
  
  comb->SetParameter(2,1000.); // Normalization
  comb->SetParameter(3,50.); // Mean = 50.
  comb->SetParameter(4,5.); // width = 5.

  gPad->SetLogy();  

  
  
  h_test_1->Fit("comb","V");
  h_test_1->Draw("e1");
  
  
  
  TestCanvas->Update();
  
  
  
   chi2 = comb->GetChisquare();
  Double_t p_n[5];
  comb->GetParameters(p_n);
  
  Double_t e_n[5];
  for(Int_t i=0; i<5; i++){
    e_n[i] = comb->GetParError(i);
  }
  
      ndf = comb->GetNDF();
  cout << endl << endl;
  cout << " Output of Fit Parameters " << endl;
  cout << " ======================== " << endl << endl;   
  
  cout << " The chi^2 = " << chi2 << " for " << ndf << " degrees of freedom " << endl;
  
  cout << " The fitted constant          value = " << p_n[0] << " +/- " << e_n[0] << endl;
  cout << " The fitted slope             value = " << p_n[1] << " +/- " << e_n[1] << endl;
  cout << endl;
  cout << " The gaussian fitted constant value = " << p_n[2] << " +/- " << e_n[2] << endl;
  cout << " The fitted mean              value = " << p_n[3] << " +/- " << e_n[3] << endl;
  cout << " The fitted sigma             value = " << p_n[4] << " +/- " << e_n[4] << endl;
  
  
  backgnd->SetParameters(p_n);
  backgnd->SetLineColor(2);
  backgnd->Draw("same");
  
  GPeak->SetParameters(&p_n[2]);
  GPeak->SetLineColor(3);
  GPeak->Draw("same");
  
  TestCanvas->Update();
  
  TestCanvas->Print("Fit_Example_10.pdf");
  TestCanvas->Print("Fit_Example_10.gif");
  
  
  
  s = new char[1];
  cout << endl << " Hit <enter> to continue" << endl;
  gets(s);
  
  hfile->Close();
}
